### ISSUE TITLE
`terraform taint` Does Not Check Terraform Version

### ISSUE THREAD

**Issue by jjorissen52** [hashicorp/terraform#23049]
### Terraform Version
```
Terraform v0.12.8
+ provider.google v2.16.0
+ provider.null v2.1.2
+ provider.random v2.2.1
# and
Terraform v0.12.10
+ provider.google v2.16.0
+ provider.null v2.1.2
+ provider.random v2.2.1
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
###################
# State Management #
###################

# manage state remotely with history in case something goes horribly wrong
terraform {                                                                                                                                                                                                           
  required_version = "0.12.8"
  backend "remote" {
      hostname = "app.terraform.io"
      organization = "..."
      workspaces {
        prefix = "..."
      }
  }
}
```

### Expected Behavior
When I run `terraform taint` with a version of terrafrom incompatible with my remote state settings, I should get an error like the below:

```
Error: Unsupported Terraform Core version

  on main.tf line 7, in terraform:
   7:   required_version = "0.12.8"

This configuration does not support Terraform version 0.12.10. To proceed,
either choose another supported Terraform version or update this version
constraint. Version constraints are normally set for good reason, so updating
the constraint may lead to other errors or unexpected behavior.
```

### Actual Behavior
The terraform version was ratcheted up to v0.12.10 in Terraform Cloud (with no way to revert as far I know).

```
Error: Error loading state: state snapshot was created by Terraform v0.12.10, which is newer than current v0.12.8; upgrade to Terraform v0.12.10 or greater to work with this state
```

### Steps to Reproduce
Run the below on a project that specifies v0.12.8 using terraform v0.12.10
`terraform0.12.10 taint $resource0.12.8`

### Additional Context
Terraform Cloud is being used to manage state remotely, but everything is being executed locally.

**Comment by pkolyvas** [comment]
@jjorissen52 Thank you for taking the time to describe the issue so clearly.

We have been tracking this and similar issues where we have failed to make clear when a particular operation is run locally on a workspace / configuration which has the Terraform Cloud remote back end enabled.

`terraform taint` is one such operation which occurs locally and can, as you've discovered, upgrade your state file, rendering it incompatible with the currently selected version of Terraform in your remote workspace.

Currently we are in the planning phase of how we intend to take a stab at this, but it is something we are actively thinking about.

**Comment by alisdair** [comment]
I think this should be fixed by #23049.
